types/src/identifiers/user_handle.rs

```diff
@@ -0,0 +1,212 @@
+// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+use std::fmt;
+
+use argon2::Argon2;
+use chrono::Duration;
+use displaydoc::Display;
+use thiserror::Error;
+
+const MIN_USER_HANDLE_LENGTH: usize = 6;
+const MAX_USER_HANDLE_LENGTH: usize = 64;
+const USER_HANDLE_CHARSET: &[u8] = b"_0123456789abcdefghijklmnopqrstuvwxyz";
+
+pub const USER_HANDLE_VALIDITY_PERIOD: Duration = Duration::days(30);
+
+/// Validated plaintext user handle
+#[derive(Clone)]
+pub struct UserHandle {
+    plaintext: String,
+}
+
+impl fmt::Debug for UserHandle {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("UserHandle")
+            .field("plaintext", &"<redacted>")
+            .finish()
+    }
+}
+
+impl UserHandle {
+    pub fn new(plaintext: String) -> Result<Self, UserHandleValidationError> {
+        Self::validate(&plaintext)?;
+        Ok(Self { plaintext })
+    }
+
+    fn validate(plaintext: &str) -> Result<(), UserHandleValidationError> {
+        if plaintext.len() < MIN_USER_HANDLE_LENGTH {
+            return Err(UserHandleValidationError::TooShort);
+        }
+        if plaintext.len() > MAX_USER_HANDLE_LENGTH {
+            return Err(UserHandleValidationError::TooLong);
+        }
+        for c in plaintext.bytes() {
+            if !USER_HANDLE_CHARSET.contains(&c) {
+                return Err(UserHandleValidationError::InvalidCharacter);
+            }
+        }
+        for pair in plaintext.as_bytes().windows(2) {
+            if pair[0] == b'_' && pair[1] == b'_' {
+                return Err(UserHandleValidationError::ConsecutiveUnderscores);
+            }
+        }
+        Ok(())
+    }
+
+    pub fn hash(&self) -> Result<UserHandleHash, UserHandleHashError> {
+        let argon2 = Argon2::default();
```

Comment:
We want explicit parameters for Argon2 handle hashing. We can do this in a follow-up.